generator_gomod.go

```diff
@@ -0,0 +1,102 @@
+package dalec
+
+import (
+	"github.com/moby/buildkit/client/llb"
+	"github.com/pkg/errors"
+)
+
+const (
+	gomodCacheDir  = "/go/pkg/mod"
+	gomodGenerator = "gomod"
+)
+
+// WithModuleDeps returns an [llb.StateOption] which will add the go modules dependencies to the [llb.StateOption]'s input state.
+// It is expected that the input state a go module cache directory without any partent directories.
+func WithModuleDeps(worker llb.State, sOpt SourceOpts, src *Source, opts ...llb.ConstraintsOpt) (llb.StateOption, error) {
+	st, err := src.AsMount("gomod", sOpt, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return func(in llb.State) llb.State {
+		for _, gen := range src.Generate {
+			in = in.With(withGomod(gen, st, worker))
+		}
+		return in
+	}, nil
+}
+
+func withGomod(g *SourceGenerator, srcSt, worker llb.State) func(llb.State) llb.State {
+	return func(in llb.State) llb.State {
+		const workDir = "/work/src"
+		var srcMount llb.RunOption
+		if g.Subpath != "" {
+			srcMount = llb.AddMount(workDir, srcSt, llb.SourcePath(g.Subpath))
+		} else {
+			srcMount = llb.AddMount(workDir, srcSt)
+		}
+		return worker.Run(
+			shArgs("go mod download"),
+			llb.AddEnv("GOMODCACHE", gomodCacheDir),
+			llb.Dir(workDir),
+			srcMount,
+		).AddMount(gomodCacheDir, in)
+	}
+}
+
+func (s *Source) isGomod() bool {
+	for _, gen := range s.Generate {
+		if gen.Gomod != nil {
+			return true
+		}
+	}
+	return false
+}
+
+// HasGomods returns true if any of the sources in the spec are a go module.
+func (s *Spec) HasGomods() bool {
+	for _, src := range s.Sources {
+		if src.isGomod() {
+			return true
+		}
+	}
+	return false
+}
+
+// GomodDeps returns an [llb.StateOption] which will add the go modules dependencies to the [llb.StateOption]'s input state
+// at the provided destination path.
+//
+// If there are no go modules in the spec this is a no-op.
+func (s *Spec) GomodDeps(sOpt SourceOpts, worker llb.State, dst string, opts ...llb.ConstraintsOpt) (llb.StateOption, error) {
+	deps := llb.Scratch()
+	var hasGomod bool
+
+	if dst == "" {
+		dst = "/"
+	}
+
+	sorted := SortMapKeys(s.Sources)
+	for _, key := range sorted {
+		src := s.Sources[key]
+		if !src.isGomod() {
+			continue
+		}
+
+		opts := append(opts, ProgressGroup("Fetch go module dependencies for source: "+key))
+		hasGomod = true
+		withMods, err := WithModuleDeps(worker, sOpt, &src, opts...)
```

Comment:
I am a bit confused on how this works, so this whole PR would allow for any specified source to have in this case a gomod generator. So for a scenario that we have at least 2 sources that have gomod gen. When we retrieve these gomods do these get merged into a single directory and state for all the gomods for the different sources? or each gomod would have it's own state? The other relevant section for this question is when preparing the sources in the rpm template: `frontend/rpm/template.go`
```go
	prepareGomods := sync.OnceFunc(func() {
		if !w.Spec.HasGomods() {
			return
		}
		fmt.Fprintf(b, "mkdir -p \"%%{_builddir}/%s\"\n", gomodsName)
		fmt.Fprintf(b, "tar -C \"%%{_builddir}/%s\" -xzf \"%%{_sourcedir}/%s.tar.gz\"\n", gomodsName, gomodsName)
	})
```
the `gomodsName` is a constant so it doesn't make sense to do this more than once if the gomod modules are merged into one for all sources.

Reply:
> When we retrieve these gomods do these get merged into a single directory and state for all the gomods for the different sources?

Yes, that is what this function is doing.
The modules are stored in the state referenced by the `deps` variable, which is getting overwritten below (L90) with `deps = deps.With(withMods)`, where `withMods` is an `llb.StateOption` that modifies the state to include the deps for an individual source.

The trick there is in the `withGomod` function where it is using return value of `llb.Run(...).AddMount(gomodCachedir, in)`.
Here the return value of `AddMount` is the the content of the mount point after the `Run` has executed.